indico/modules/events/registration/google_wallet.py

```diff
@@ -0,0 +1,356 @@
+from requests.exceptions import HTTPError, RequestException
+from werkzeug.exceptions import ServiceUnavailable
+
+from indico.core import signals
+from indico.core.config import config
+from indico.core.logger import Logger
+from indico.modules.events.registration.settings import event_wallet_settings
+from indico.util.i18n import _
+from indico.web.flask.util import url_for
+
+
+API_BASE_URL = 'https://walletobjects.googleapis.com/walletobjects/v1'
+API_CLASS_URL = f'{API_BASE_URL}/eventTicketClass'
+API_OBJECT_URL = f'{API_BASE_URL}/eventTicketObject'
+logger = Logger.get('events.registration.google_wallet')
+
+
+class GoogleCredentialValidationResult(Enum):
+    ok = auto()
+    invalid = auto()
+    refused = auto()
+    bad_issuer = auto()
+    failed = auto()
+
+
+def handle_http_errors(*, silent=False):
+    def decorator(fn):
+        @wraps(fn)
+        def wrapper(*args, **kwargs):
+            try:
+                return fn(*args, **kwargs)
+            except HTTPError as exc:
+                resp = exc.response
+                logger.warning('Google Wallet API request (%s) failed [%d]: %s',
+                               fn.__name__, resp.status_code, resp.text)
+                if not silent:
+                    raise ServiceUnavailable(_('Could not generate ticket')) from exc
+            except RequestException as exc:
+                logger.warning('Google Wallet API request (%s) failed: %s', fn.__name__, exc)
+                if not silent:
+                    raise ServiceUnavailable(_('Could not generate ticket')) from exc
+        return wrapper
+    return decorator
+
+
+class GoogleWalletManager:
+    """Google Wallet ticketing manager.
+
+    This class handles the integration with the Google Wallet ticketing API.
+    """
+
+    def __init__(self, event):
+        self.event = event
+        self.settings = self.get_google_wallet_settings(self.event.category)
+        self.credentials, self.http_client = self._load_credentials()
+
+    def _load_credentials(self):
+        """Load Google credentials and setup authenticated HTTP client."""
+        if not self.configured:
+            return None, None
+        credentials = Credentials.from_service_account_info(
+            self.settings['google_wallet_credentials'],
+            scopes=['https://www.googleapis.com/auth/wallet_object.issuer']
+        )
+        http_client = AuthorizedSession(credentials)
+        return credentials, http_client
+
+    @property
+    def configured(self):
+        return config.ENABLE_GOOGLE_WALLET and bool(self.settings['google_wallet_mode'])
+
+    @property
+    def class_id(self):
+        """The unique identifier for an event ticket class.
+
+        Each event has one ticket class that holds shared data such as the title
+        and dates of the event.
+        """
+        return event_wallet_settings.get(self.event, 'google_wallet_class_id', f'Indico-Event-{self.event.id}')
```

Comment:
It is not configurable. It used to be fully hardcoded, I just started storing it in the DB because that way we know whether a ticket class has been created for an event so we don't need to start making Google API calls on any update as soon as the regform has GW ticketing enabled.

Another benefit of storing it is that - in case it's ever needed - we could quickly query the DB to see which events actively use GW (e.g. if we ever need to bulk update all ticket classes).

Reply:
FWIW the main reason why I don't store it for the registration as well is that I'm quite sure that at some point we'll need to extend this whole thing to support generating Google Wallet tickets for accompanying persons as well.... and we don't have a "RegistrationSettingsProxy", so it'd require a full column for the ticket ID, and probably a JSONB column for accompanying persons (mapping their UUID to the ticket ID)...